comp/remote-config/rcclient/rcclient.go

```diff
@@ -282,3 +293,85 @@ type ListenerProvider struct {
 
 	ListenerProvider RCListener `group:"rCListener"`
 }
+
+func (rc rcClient) onAPMTracingUpdate(update map[string]state.RawConfig, applyStateCallback func(string, state.ApplyStatus)) { //nolint:revive
+	if len(update) == 0 {
+		// Empty update means revert to default behavior, so remove any existing config file
+		err := os.Remove(apmTracingFilePath)
+		if err == nil {
+			pkglog.Infof("Removed APM_TRACING remote config file, APM injection will revert to default behavior")
+		} else if !os.IsNotExist(err) {
+			// If the file already wasn't there then it wasn't an error
+			pkglog.Errorf("Failed to remove APM_TRACING remote config file, previous APM injection behavior will continue")
+		}
+		return
+	}
+
+	var senvConfigs []serviceEnvConfig
+	// Maps update IDs to their error, empty string indicates success
+	updateStatus := map[string]string{}
+	var hostTracingEnabled bool
+	var hostEnvTarget string
+	for id, rawConfig := range update {
+		tcu := tracingConfigUpdate{}
+		err := json.Unmarshal(rawConfig.Config, &tcu)
+		updateStatus[id] = ""
+		if err != nil {
+			pkglog.Warnf("Skipping invalid APM_TRACING remote update %s: %v, any err: %v", id, tcu, err)
+			updateStatus[id] = InvalidAPMTracingPayload
+			continue
+		}
+		pkglog.Infof("Received APM_TRACING remote update %s: %v, any err: %v", id, tcu, err)
+		if tcu.InfraTarget != nil {
+			// This is an infra targeting payload, skip adding it to the service env config map
+			hostTracingEnabled = tcu.LibConfig.TracingEnabled
+			hostEnvTarget = tcu.LibConfig.Env
+			continue
+		}
+		if tcu.ServiceTarget == nil {
+			pkglog.Warnf("Missing service_target from APM_TRACING config update, SKIPPING: %v", tcu)
+			updateStatus[id] = MissingServiceTarget
+			continue
+		}
+		senvConfigs = append(senvConfigs, serviceEnvConfig{
+			Service:        tcu.ServiceTarget.Service,
+			Env:            tcu.ServiceTarget.Env,
+			TracingEnabled: tcu.LibConfig.TracingEnabled,
+		})
+	}
+	tec := tracingEnabledConfig{
+		TracingEnabled:    hostTracingEnabled,
+		Env:               hostEnvTarget,
+		ServiceEnvConfigs: senvConfigs,
+	}
+	configFile, err := yamlv2.Marshal(tec)
+	if err != nil {
+		pkglog.Errorf("Failed to marshal APM_TRACING config update %v", err)
+		return
+	}
```

Comment:
Shouldn't it be reported as an error via `applyStateCallback` ?

Reply:
Maybe? this is a weird one since I think this error is impossible at compile time? 🤔

Reply:
The role of `applyStateCallback` is to report errors to the RC platform and eventually to the user on the UI, right? My understanding was that any error should be reported.

Either way, this is a non-blocking comment, I was just curious why this error was handled differently.